pkg/observer/polling/observer.go

```diff
@@ -0,0 +1,337 @@
+/*
+A polling observer is an Observer that continuously polls every block, samples
+available upkeeps, and surfaces upkeeps that should be agreed on by a network.
+*/
+package polling
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/ocr2keepers/internal/util"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+	"github.com/smartcontractkit/ocr2keepers/pkg/observer"
+)
+
+var ErrTooManyErrors = fmt.Errorf("too many errors in parallel worker process")
+
+// UpkeepProvider is a dependency used by the observer to get upkeeps
+// available for polling
+type UpkeepProvider interface {
+	GetActiveUpkeepIDs(context.Context) ([]ocr2keepers.UpkeepIdentifier, error)
+}
+
+// Encoder is a dependency that provides helper functions for results and keys
+type Encoder interface {
+	// Eligible returns whether or not the upkeep result should be performed
+	Eligible(ocr2keepers.UpkeepResult) (bool, error)
+	// MakeUpkeepKey combines a block and upkeep id into an upkeep key. This
+	// will probably go away with a more structured static upkeep type.
+	MakeUpkeepKey(ocr2keepers.BlockKey, ocr2keepers.UpkeepIdentifier) ocr2keepers.UpkeepKey
+	// SplitUpkeepKey ...
+	SplitUpkeepKey(ocr2keepers.UpkeepKey) (ocr2keepers.BlockKey, ocr2keepers.UpkeepIdentifier, error)
+	// Detail is a temporary value that provides upkeep key and gas to perform.
+	// A better approach might be needed here.
+	Detail(ocr2keepers.UpkeepResult) (ocr2keepers.UpkeepKey, uint32, error)
```

Comment:
Not fully understand the need here. As I see it's only being used to fetch the upkeepKey from the result currently?

Reply:
if `ocr2keepers.UpkeepResult` was better typed here instead of interface then we wouldn't need this I think

Reply:
This will be resolved in the next version